pkg/reconciler/delivery/policy.go

```diff
@@ -52,29 +54,65 @@ type Stage struct {
 // computeNewPercent calculates, given a Policy and the current rollout stage,
 // the traffic percentage for the NEW Revision in the next rollout stage
 func computeNewPercent(p *Policy, currentPercent int) (int, error) {
-	i := sort.Search(len(p.Percents), func(i int) bool {
-		return p.Percents[i].Percent >= currentPercent
+	i := sort.Search(len(p.Stages), func(i int) bool {
+		return p.Stages[i].Percent >= currentPercent
 	})
-	if i < len(p.Percents) && p.Percents[i].Percent == currentPercent {
-		if i == len(p.Percents) - 1 {
+	if i < len(p.Stages) && p.Stages[i].Percent == currentPercent {
+		if i == len(p.Stages) - 1 {
 			return 100, nil
 		}
-		return p.Percents[i + 1].Percent, nil
+		return p.Stages[i + 1].Percent, nil
 	}
 	return 0, fmt.Errorf("invalid percentage for current rollout stage")
 }
 
 // getThreshold returns, given the percentage for a rollout stage, its corresponding threshold value
 // if the threshold value isn't specified, DefaultThreshold is used
 func getThreshold(p *Policy, currentPercent int) (int, error) {
-	i := sort.Search(len(p.Percents), func(i int) bool {
-		return p.Percents[i].Percent >= currentPercent
+	i := sort.Search(len(p.Stages), func(i int) bool {
+		return p.Stages[i].Percent >= currentPercent
 	})
-	if i < len(p.Percents) && p.Percents[i].Percent == currentPercent {
-		if p.Percents[i].Threshold != nil {
-			return *p.Percents[i].Threshold, nil
+	if i < len(p.Stages) && p.Stages[i].Percent == currentPercent {
+		if p.Stages[i].Threshold != nil {
+			return *p.Stages[i].Threshold, nil
 		}
 		return p.DefaultThreshold, nil
 	}
 	return 0, fmt.Errorf("invalid percentage for current rollout stage")
 }
+
+// computeNewPercentExplicit is an explicit way of computing a percentage without relying on the previous stage
+// elapsed is the total time duration since the beginning of the rollout
+// this function doesn't return an error because an error is impossible
+func computeNewPercentExplicit(p *Policy, elapsed time.Duration) int {
+	metric := float64(elapsed) / float64(time.Second)
+	metricCumulative := 0
+	for _, s := range p.Stages[1:] {
+		extra := p.DefaultThreshold
+		if s.Threshold != nil {
+			extra = *s.Threshold
+		}
+		metricCumulative += extra
+		if float64(metricCumulative) > metric {
+			return s.Percent
+		}
+	}
+	return 100
+}
+
+// metricTillNextStage computes how much time (full seconds) to wait before progressing to the next stage
+func metricTillNextStage(p *Policy, elapsed time.Duration) int {
+	metric := float64(elapsed) / float64(time.Second)
+	metricCumulative := 0
+	for _, s := range p.Stages[1:] {
+		extra := p.DefaultThreshold
+		if s.Threshold != nil {
+			extra = *s.Threshold
+		}
+		metricCumulative += extra
+		if float64(metricCumulative) > metric {
+			return int(float64(metricCumulative) - metric) + 1 // +1 deals with float truncating
```

Comment:
That's not correct. You should probably use math.Round then.

Reply:
I used `+1` here on purpose. Using `math.Round` will cause incorrect behavior later on. For example, if `float64(metricCumulative) - metric)` is 2.4 seconds, math.Round will return 2. However, if the event is enqueued 2 seconds later instead of 2.4 seconds later, then when that event actually triggers the reconciler, the reconciler will still think that the timestamp hasn't expired (short by 0.4 seconds), so it wouldn't update the Route even when it should. I used `+1` here to make sure that `metricTillNextStage` always returns a result that is strictly bigger than the de facto result to prevent such incorrect behavior.

Reply:
Then the comment is not correct :)